File: src-tauri/src/image_commands/path_utils.rs
use md5;
use std::path::{Path, PathBuf};
use tauri::{AppHandle, Manager};

pub fn normalize_extension(raw_ext: &str) -> String {
    let ext = raw_ext.trim().trim_start_matches('.').to_ascii_lowercase();
    if ext.is_empty() {
        return "png".to_string();
    }

    if ext == "jpeg" {
        return "jpg".to_string();
    }

    ext
}

pub fn extension_from_mime(mime: &str) -> String {
    let normalized = mime.trim().to_ascii_lowercase();
    match normalized.as_str() {
        "image/png" => "png".to_string(),
        "image/jpeg" => "jpg".to_string(),
        "image/jpg" => "jpg".to_string(),
        "image/webp" => "webp".to_string(),
        "image/gif" => "gif".to_string(),
        "image/bmp" => "bmp".to_string(),
        "image/avif" => "avif".to_string(),
        _ => "png".to_string(),
    }
}

pub fn mime_from_extension(ext: &str) -> &'static str {
    match normalize_extension(ext).as_str() {
        "png" => "image/png",
        "jpg" | "jpeg" => "image/jpeg",
        "webp" => "image/webp",
        "gif" => "image/gif",
        "bmp" => "image/bmp",
        "avif" => "image/avif",
        _ => "application/octet-stream",
    }
}

pub fn resolve_data_root_dir(app: &AppHandle) -> Result<PathBuf, String> {
    let app_local = app
        .path()
        .app_local_data_dir()
        .map_err(|e| format!("Failed to resolve app local data dir: {}", e))?;

    let root = app_local.join("Henji-AI");
    std::fs::create_dir_all(&root).map_err(|e| format!("Failed to create data root dir: {}", e))?;
    Ok(root)
}

pub fn resolve_uploads_dir(app: &AppHandle) -> Result<PathBuf, String> {
    let uploads_dir = resolve_data_root_dir(app)?.join("Uploads");
    std::fs::create_dir_all(&uploads_dir)
        .map_err(|e| format!("Failed to create uploads dir: {}", e))?;
    Ok(uploads_dir)
}

pub fn resolve_debug_dir(app: &AppHandle, category: &str) -> Result<PathBuf, String> {
    let normalized_category = sanitize_file_stem(category);
    let debug_dir = resolve_data_root_dir(app)?
        .join("debug")
        .join(normalized_category);
    std::fs::create_dir_all(&debug_dir)
        .map_err(|e| format!("Failed to create debug dir: {}", e))?;
    Ok(debug_dir)
}

pub fn persist_image_bytes(
    app: &AppHandle,
    bytes: &[u8],
    extension: &str,
) -> Result<String, String> {
    if bytes.is_empty() {
        return Err("Image bytes are empty".to_string());
    }

    let digest = md5::compute(bytes);
    let ext = normalize_extension(extension);
    let filename = format!("{:x}.{}", digest, ext);
    let output_path = resolve_uploads_dir(app)?.join(filename);

    if !output_path.exists() {
        std::fs::write(&output_path, bytes)
            .map_err(|e| format!("Failed to persist generated image: {}", e))?;
    }

    Ok(output_path.to_string_lossy().to_string())
}

pub fn sanitize_file_stem(raw: &str) -> String {
    let trimmed = raw.trim();
    let fallback = "storyboard-image";
    if trimmed.is_empty() {
        return fallback.to_string();
    }

    let mut sanitized = String::with_capacity(trimmed.len());
    for ch in trimmed.chars() {
        let blocked = matches!(ch, '<' | '>' | ':' | '"' | '/' | '\\' | '|' | '?' | '*');
        if blocked || ch.is_control() {
            continue;
        }
        sanitized.push(ch);
    }

    let compact = sanitized.trim().trim_matches('.').to_string();
    if compact.is_empty() {
        fallback.to_string()
    } else {
        compact
    }
}

pub fn ensure_unique_path(path: PathBuf) -> PathBuf {
    if !path.exists() {
        return path;
    }

    let parent = path
        .parent()
        .map(Path::to_path_buf)
        .unwrap_or_else(|| PathBuf::from("."));
    let stem = path
        .file_stem()
        .and_then(|value| value.to_str())
        .unwrap_or("storyboard-image");
    let ext = path
        .extension()
        .and_then(|value| value.to_str())
        .unwrap_or("png");

    for index in 1..10_000_u32 {
        let candidate = parent.join(format!("{}-{}.{}", stem, index, ext));
        if !candidate.exists() {
            return candidate;
        }
    }

    path
}

pub fn ensure_output_path_with_extension(path: &Path, extension: &str) -> PathBuf {
    if path.extension().is_some() {
        return path.to_path_buf();
    }

    let mut with_extension = path.to_path_buf();
    with_extension.set_extension(normalize_extension(extension));
    with_extension
}

File: electron/main/services/image/path-utils.ts
import { app } from 'electron'
import crypto from 'node:crypto'
import fs from 'node:fs'
import path from 'node:path'
import { getCustomDataRoot } from '../dataRoot'
import { releaseManagedMediaFileLease } from './managed-media-leases'

const APP_IDENTIFIER = 'com.henji.ai'
const uploadImageLeaseCounts = new Map<string, number>()

export function normalizeExtension(rawExt: string | undefined): string {
  const ext = (rawExt ?? '').trim().replace(/^\./, '').toLowerCase()
  if (!ext) return 'png'
  return ext === 'jpeg' ? 'jpg' : ext
}

export function extensionFromMime(mime: string | undefined): string {
  const normalized = (mime ?? '').trim().toLowerCase().split(';')[0]
  if (normalized === 'image/png') return 'png'
  if (normalized === 'image/jpeg' || normalized === 'image/jpg') return 'jpg'
  if (normalized === 'image/webp') return 'webp'
  if (normalized === 'image/gif') return 'gif'
  if (normalized === 'image/bmp') return 'bmp'
  if (normalized === 'image/avif') return 'avif'
  return 'png'
}

export function mimeFromExtension(extension: string | undefined): string {
  switch (normalizeExtension(extension)) {
    case 'png': return 'image/png'
    case 'jpg': return 'image/jpeg'
    case 'webp': return 'image/webp'
    case 'gif': return 'image/gif'
    case 'bmp': return 'image/bmp'
    case 'avif': return 'image/avif'
    default: return 'application/octet-stream'
  }
}

export function getDataRootDir(): string {
  const custom = getCustomDataRoot()
  const root = custom
    ? custom
    : path.join(
        process.platform === 'win32' && process.env.LOCALAPPDATA
          ? path.join(process.env.LOCALAPPDATA, APP_IDENTIFIER)
          : path.join(app.getPath('appData'), APP_IDENTIFIER),
        'Henji-AI'
      )
  fs.mkdirSync(root, { recursive: true })
  return root
}

export function getUploadsDir(): string {
  const uploads = path.join(getDataRootDir(), 'Uploads')
  fs.mkdirSync(uploads, { recursive: true })
  return uploads
}

export function getDebugDir(category: string): string {
  const debugDir = path.join(getDataRootDir(), 'debug', sanitizeFileStem(category))
  fs.mkdirSync(debugDir, { recursive: true })
  return debugDir
}

export function persistImageBytes(bytes: Buffer, extension: string | undefined): string {
  return persistContentAddressedImage(bytes, extension).filePath
}

export interface PersistedImageBytes {
  filePath: string
  /** true 表示调用方持有一个可释放 lease，不再只表示本次首次创建了文件。 */
  created: boolean
}

interface ContentAddressedImageResult {
  filePath: string
  created: boolean
}

function persistContentAddressedImage(
  bytes: Buffer,
  extension: string | undefined,
): ContentAddressedImageResult {
  if (bytes.length === 0) throw new Error('Image bytes are empty')
  const digest = crypto.createHash('md5').update(bytes).digest('hex')
  const ext = normalizeExtension(extension)
  const filePath = path.join(getUploadsDir(), `${digest}.${ext}`)
  const created = !fs.existsSync(filePath)
  if (created) fs.writeFileSync(filePath, bytes)
  return { filePath, created }
}

function releaseUploadImageLease(filePath: string): void {
  const leaseCount = uploadImageLeaseCounts.get(filePath)
  if (leaseCount === undefined) {
    fs.rmSync(filePath, { force: true })
    return
  }
  if (leaseCount > 1) {
    uploadImageLeaseCounts.set(filePath, leaseCount - 1)
    return
  }
  uploadImageLeaseCounts.delete(filePath)
  fs.rmSync(filePath, { force: true })
}

/**
 * 供需要失败回滚的批处理使用。相同内容在同一进程内并行持有独立 lease，
 * 只有最后一个 lease 释放时才删除文件；进程启动前已存在的文件不取得所有权。
 */
export function persistImageBytesTracked(bytes: Buffer, extension: string | undefined): PersistedImageBytes {
  const persisted = persistContentAddressedImage(bytes, extension)
  const filePath = path.resolve(persisted.filePath)
  const currentLeaseCount = uploadImageLeaseCounts.get(filePath) ?? 0
  if (!persisted.created && currentLeaseCount === 0) {
    return { filePath, created: false }
  }
  uploadImageLeaseCounts.set(filePath, currentLeaseCount + 1)
  return { filePath, created: true }
}

export function rollbackPersistedImageBytes(entry: PersistedImageBytes): void {
  if (!entry.created) return
  try {
    releaseUploadImageLease(path.resolve(entry.filePath))
  } catch {
    // 失败回滚属于 best-effort；原始错误必须继续上抛。
  }
}

export function releaseManagedImagePaths(filePaths: readonly string[]): void {
  const uploadsDir = path.resolve(getUploadsDir())
  const prefix = `${uploadsDir}${path.sep}`
  for (const filePath of new Set(filePaths)) {
    const resolved = path.resolve(filePath)
    if (!resolved.startsWith(prefix)) throw new Error('只能释放 Uploads 目录内的受管图片')
    releaseUploadImageLease(resolved)
  }
}

/** 通用生成事务回滚入口；只接受应用数据根内的 Uploads/Media 受管文件。 */
export function releaseManagedGenerationMediaPaths(filePaths: readonly string[]): void {
  const uploadsRoot = path.resolve(getDataRootDir(), 'Uploads')
  const mediaRoot = path.resolve(getDataRootDir(), 'Media')
  for (const filePath of new Set(filePaths)) {
    const resolved = path.resolve(filePath)
    const isUpload = resolved.startsWith(`${uploadsRoot}${path.sep}`)
    const isMedia = resolved.startsWith(`${mediaRoot}${path.sep}`)
    if (!isUpload && !isMedia) {
      throw new Error('只能释放应用数据目录内的受管生成媒体')
    }
    if (isUpload) {
      releaseUploadImageLease(resolved)
      continue
    }
    if (isMedia && releaseManagedMediaFileLease(resolved) === 'retained') continue
    fs.rmSync(resolved, { force: true })
  }
}

export function sanitizeFileStem(raw: string | undefined): string {
  const trimmed = (raw ?? '').trim()
  if (!trimmed) return 'storyboard-image'
  const sanitized = Array.from(trimmed)
    .filter((char) => !isBlockedFilenameChar(char))
    .join('')
    .trim()
    .replace(/^\.+|\.+$/g, '')
  return sanitized || 'storyboard-image'
}

function isBlockedFilenameChar(char: string): boolean {
  return char.charCodeAt(0) < 32 || '<>:"/\\|?*'.includes(char)
}

export function ensureUniquePath(targetPath: string): string {
  if (!fs.existsSync(targetPath)) return targetPath
  const parsed = path.parse(targetPath)
  for (let index = 1; index < 10000; index += 1) {
    const candidate = path.join(parsed.dir, `${parsed.name}-${index}${parsed.ext}`)
    if (!fs.existsSync(candidate)) return candidate
  }
  return targetPath
}

export function ensureOutputPathWithExtension(targetPath: string, extension: string): string {
  return path.extname(targetPath) ? targetPath : `${targetPath}.${normalizeExtension(extension)}`
}

export function writeBytesToPath(targetPath: string, bytes: Buffer): void {
  fs.mkdirSync(path.dirname(targetPath), { recursive: true })
  fs.writeFileSync(targetPath, bytes)
}
